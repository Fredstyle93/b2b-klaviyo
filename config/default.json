{
  "order": {
    "states": {
      "changed": {
        "cancelledOrder": [
          "Cancelled"
        ],
        "fulfilledOrder": [
          "Confirmed",
          "Complete"
        ]
      },
      "created": {
        "placedOrder": [
          "Open"
        ]
      }
    },
    "messages": {
      "changed": [
        "OrderStateChanged"
      ],
      "created": [
        "OrderCreated",
        "OrderImported",
        "OrderCustomerSet"
      ]
    },
    "properties": {
      "include": [
        "customerId",
        "customerEmail",
        "id",
        "createdAt",
        "lastModifiedAt",
        "lineItems",
        "customLineItems",
        "totalPrice",
        "orderState",
        "version"
      ],
      "exclude": [
        "paymentInfo",
        "shippingMode",
        "syncInfo",
        "origin",
        "refusedGifts",
        "directDiscounts",
        "discountCodes",
        "transactionFee",
        "createdBy"
      ],
      "map": {
        "id": "orderId",
        "lineItems": "items",
        "customLineItems": "customItems"
      }
    },
    "metrics": {
      "cancelledOrder": "Cancelled Order",
      "fulfilledOrder": "Fulfilled Order",
      "placedOrder": "Placed Order",
      "orderedProduct": "Ordered Product",
      "refundedOrder": "Refunded Order"
    }
  },
  "customer": {
    "messages": {
      "addressChanged": [
        "CustomerAddressAdded",
        "CustomerAddressRemoved",
        "CustomerAddressChanged"
      ]
    }
  },
  "payment": {
    "messages": {
      "transactionAdded": [
        "PaymentTransactionAdded"
      ],
      "transactionChanged": [
        "PaymentTransactionStateChanged"
      ]
    },
    "states": {
      "validTransactionStates": [
        "Initial",
        "Pending",
        "Success"
      ]
    }
  }
}