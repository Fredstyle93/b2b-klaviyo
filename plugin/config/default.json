{
  "disabledEvents": [],
  "order": {
    "states": {
      "changed": {
        "readyToBePickedUp": [
          "Confirmed"
        ],
        "cancelledOrder": [
          "Cancelled"
        ],
        "fulfilledOrder": [
          "Complete"
        ]
      },
      "created": {
        "placedOrder": [
          "Open"
        ]
      }
    },
    "messages": {
      "changed": [
        "OrderStateChanged"
      ],
      "created": [
        "OrderCreated",
        "OrderImported",
        "OrderCustomerSet"
      ]
    },
    "properties": {
      "include": [],
      "exclude": [],
      "map": {}
    },
    "metrics": {
      "readyToBePickedUp": "Ready To Pick Up Order",
      "cancelledOrder": "Cancelled Order",
      "fulfilledOrder": "Fulfilled Order",
      "placedOrder": "Placed Order",
      "orderedProduct": "Ordered Product",
      "refundedOrder": "Refunded Order"
    },
    "customFields": {
      "properties": {
        "include": [],
        "exclude": [],
        "map": {}
      }
    }
  },
  "customer": {
    "messages": {
      "customerCreated": [
        "CustomerCreated"
      ],
      "addressChanged": [
        "CustomerAddressAdded",
        "CustomerAddressRemoved",
        "CustomerAddressChanged"
      ]
    },
    "customFields": {
      "properties": {
        "include": [],
        "exclude": [],
        "map": {}
      }
    },
    "metrics": {
      "customerCreated": "Customer Created"
    }
  },
  "payment": {
    "messages": {
      "transactionAdded": [
        "PaymentTransactionAdded"
      ],
      "transactionChanged": [
        "PaymentTransactionStateChanged"
      ]
    },
    "states": {
      "validTransactionStates": [
        "Initial",
        "Pending",
        "Success"
      ]
    }
  },
  "category": {
    "messages": {
      "created": [
        "CategoryCreated"
      ]
    }
  },
  "product": {
    "messages": {
      "deleted": [
        "ProductUnpublished"
      ],
      "published": [
        "ProductPublished"
      ]
    },
    "inventory": {
      "useChannelInventory": ""
    }
  }
}
