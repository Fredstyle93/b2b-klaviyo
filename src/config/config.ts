const config: any = {
    order: {
        changedStates: [
            {
                sources: ['Cancelled'],
                value: 'Cancelled Order',
            },
            {
                sources: ['Confirmed', 'Complete'],
                value: 'Fulfilled Order',
            },
        ],
        createdStates: [
            {
                sources: ['OrderCreated', 'Open'],
                value: 'Order created',
            },
            {
                sources: ['OrderedProduct'],
                value: 'Ordered Product',
            },
        ],
        refundedStates: [
            {
                sources: ['OrderRefunded'],
                value: 'Refunded Order',
            },
        ],
        messageTypes: [
            {
                sources: ['OrderCreated', 'OrderImported', 'OrderCustomerSet'],
                value: true,
            },
        ],
        allowedProperties: {
            customerId: null,
            customerEmail: null,
            id: 'orderId',
            createdAt: null,
            lastModifiedAt: null,
            lineItems: 'items',
            customLineItems: 'customItems',
            totalPrice: null,
            orderState: null,
        },
    },
    customer: {
        addressChangeMessages: [
            {
                sources: ['CustomerAddressAdded', 'CustomerAddressRemoved', 'CustomerAddressChanged'],
                value: true,
            },
        ],
    },
};

export default config;
