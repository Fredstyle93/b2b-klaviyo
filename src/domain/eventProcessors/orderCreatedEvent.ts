import { AbstractEvent } from './abstractEvent';
import logger from '../../utils/log';
import { OrderCreatedMessage } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/message';
import { Order, OrderState } from '@commercetools/platform-sdk';
import { getTypedMoneyAsNumber } from '../../utils/get-typed-money-as-number';
import { getConfigProperty } from '../../utils/prop-mapper';
import { getCustomerProfileFromOrder } from '../../utils/get-customer-profile-from-order';

export class OrderCreatedEvent extends AbstractEvent {
    isEventValid(): boolean {
        const orderCreatedMessage = this.ctMessage as unknown as OrderCreatedMessage;
        return (
            orderCreatedMessage.resource.typeId === 'order' &&
            orderCreatedMessage.type === 'OrderCreated' &&
            !!orderCreatedMessage.order &&
            (!!orderCreatedMessage.order.customerEmail || !!orderCreatedMessage.order.customerId) &&
            this.isValidState(orderCreatedMessage.order?.orderState)
        );
    }

    generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const orderCreatedMessage = this.ctMessage as unknown as OrderCreatedMessage;
        logger.info('Processing order created event');

        const body = {
            data: {
                type: 'event',
                attributes: {
                    profile: getCustomerProfileFromOrder(orderCreatedMessage.order),
                    metric: {
                        name: this.getOrderMetric('OrderCreated'),
                    },
                    value: getTypedMoneyAsNumber(orderCreatedMessage.order?.totalPrice),
                    properties: { ...orderCreatedMessage.order } as any,
                    unique_id: orderCreatedMessage.order.id,
                    time: orderCreatedMessage.order.createdAt,
                },
            },
        };

        const events: KlaviyoEvent[] = [{ body, type: 'event' }];

        this.getProductOrderedEventsFromOrder(events, orderCreatedMessage.order);

        return Promise.resolve(events);
    }

    private getCustomerProfile(order: Order): KlaviyoEventProfile {
        const profile: KlaviyoEventProfile = {};
        if (order.customerEmail) {
            profile.$email = order.customerEmail;
        }
        if (order.customerId) {
            profile.$id = order.customerId;
        }
        return profile;
    }

    private isValidState(orderState: OrderState): boolean {
        return Boolean(getConfigProperty('order.createdStates', orderState));
    }

    private getProductOrderedEventsFromOrder(events: KlaviyoEvent[], order: Order) {
        order?.lineItems?.forEach((lineItem) => {
            events.push({
                body: {
                    data: {
                        type: 'event',
                        attributes: {
                            profile: this.getCustomerProfile(order),
                            metric: {
                                name: this.getOrderMetric('OrderedProduct'),
                            },
                            value: getTypedMoneyAsNumber(lineItem.totalPrice),
                            properties: { ...lineItem },
                            unique_id: lineItem.id,
                            time: order.createdAt,
                        },
                    },
                },
                type: 'event',
            });
        });
    }

    private getOrderMetric(type: string): string {
        return getConfigProperty('order.createdStates', type);
    }
}
