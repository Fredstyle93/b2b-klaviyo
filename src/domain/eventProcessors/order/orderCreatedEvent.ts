import { AbstractEvent } from '../abstractEvent';
import logger from '../../../utils/log';
import { Order, OrderCreatedMessage, OrderCustomerSetMessage, OrderState } from '@commercetools/platform-sdk';
import { getTypedMoneyAsNumber } from '../../../utils/get-typed-money-as-number';
import { getCustomerProfileFromOrder } from '../../../utils/get-customer-profile-from-order';
import { getOrderById } from '../../ctService';
import { mapAllowedProperties } from '../../../utils/property-mapper';
import config from 'config';

export class OrderCreatedEvent extends AbstractEvent {
    isEventValid(): boolean {
        const message = this.ctMessage as unknown as OrderCreatedMessage | OrderCustomerSetMessage;
        return (
            message.resource.typeId === 'order' &&
            this.isValidMessageType(message.type) &&
            this.hasExpectedMessageProperties(message)
        );
    }

    async generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const message = this.ctMessage as unknown as OrderCreatedMessage | OrderCustomerSetMessage;
        logger.info('Processing order created event');

        let order: Order;
        if ('order' in message) {
            order = message.order;
        } else {
            order = (await getOrderById(message.resource.id)) as Order;
        }

        const body: EventRequest = {
            data: {
                type: 'event',
                attributes: {
                    profile: getCustomerProfileFromOrder(order),
                    metric: {
                        name: config.get('order.metrics.placedOrder'),
                    },
                    value: this.context.currencyService.convert(
                        getTypedMoneyAsNumber(order?.totalPrice),
                        order.totalPrice.currencyCode,
                    ),
                    properties: mapAllowedProperties('order', { ...order }) as any,
                    unique_id: order.id,
                    time: order.createdAt,
                },
            },
        };

        const events: KlaviyoEvent[] = [{ body, type: 'event' }];

        this.getProductOrderedEventsFromOrder(events, order);

        return Promise.resolve(events);
    }

    private getCustomerProfile(order: Order): KlaviyoEventProfile {
        const profile: KlaviyoEventProfile = {};
        if (order.customerEmail) {
            profile.$email = order.customerEmail;
        }
        if (order.customerId) {
            profile.$id = order.customerId;
        }
        return profile;
    }

    private getProductOrderedEventsFromOrder(events: KlaviyoEvent[], order: Order) {
        order?.lineItems?.forEach((lineItem) => {
            events.push({
                body: {
                    data: {
                        type: 'event',
                        attributes: {
                            profile: this.getCustomerProfile(order),
                            metric: {
                                name: config.get('order.metrics.orderedProduct'),
                            },
                            value: getTypedMoneyAsNumber(lineItem.totalPrice),
                            properties: { ...lineItem },
                            unique_id: lineItem.id,
                            time: order.createdAt,
                        },
                    },
                },
                type: 'event',
            });
        });
    }

    private isValidState(orderState: OrderState): boolean {
        return Boolean(
            config.has('order.states.created') &&
                (config.get('order.states.created.placedOrder') as string[])?.includes(orderState),
        );
    }

    private isValidMessageType(type: string): boolean {
        return Boolean(
            config.has('order.messages.created') && (config.get('order.messages.created') as string[])?.includes(type),
        );
    }

    private hasExpectedMessageProperties(message: OrderCreatedMessage | OrderCustomerSetMessage) {
        return (
            !!(message as OrderCustomerSetMessage).customer ||
            (!!(message as OrderCreatedMessage).order &&
                (!!(message as OrderCreatedMessage).order?.customerEmail ||
                    !!(message as OrderCreatedMessage).order?.customerId) &&
                this.isValidState((message as OrderCreatedMessage).order?.orderState))
        );
    }
}
