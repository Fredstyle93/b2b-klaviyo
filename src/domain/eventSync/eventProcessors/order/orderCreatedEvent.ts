import { AbstractEventProcessor } from '../abstractEventProcessor';
import logger from '../../../../utils/log';
import { Order, OrderCreatedMessage, OrderCustomerSetMessage, OrderState } from '@commercetools/platform-sdk';
import { getTypedMoneyAsNumber } from '../../../../utils/get-typed-money-as-number';
import { getCustomerProfileFromOrder } from '../../../../utils/get-customer-profile-from-order';
import { getOrderById } from '../../../../infrastructure/driven/commercetools/ctService';
import { mapAllowedProperties } from '../../../../utils/property-mapper';
import config from 'config';

export class OrderCreatedEvent extends AbstractEventProcessor {
    isEventValid(): boolean {
        const message = this.ctMessage as unknown as OrderCreatedMessage | OrderCustomerSetMessage;
        return (
            message.resource.typeId === 'order' &&
            this.isValidMessageType(message.type) &&
            this.hasExpectedMessageProperties(message)
        );
    }

    async generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const message = this.ctMessage as unknown as OrderCreatedMessage | OrderCustomerSetMessage;
        logger.info('Processing order created event');

        let order: Order;
        if ('order' in message) {
            order = message.order;
        } else {
            order = (await getOrderById(message.resource.id)) as Order;
        }

        const body: EventRequest = this.context.orderMapper.mapCtOrderToKlaviyoEvent(order, config.get('order.metrics.placedOrder'));

        const events: KlaviyoEvent[] = [{ body, type: 'event' }];

        this.getProductOrderedEventsFromOrder(events, order);

        return Promise.resolve(events);
    }

    private getProductOrderedEventsFromOrder(events: KlaviyoEvent[], order: Order) {
        order?.lineItems?.forEach((lineItem) => {
            events.push({
                body: this.context.orderMapper.mapOrderLineToProductOrderedEvent(lineItem, order),
                type: 'event',
            });
        });
    }

    private isValidState(orderState: OrderState): boolean {
        return Boolean(
            config.has('order.states.created') &&
                (config.get('order.states.created.placedOrder') as string[])?.includes(orderState),
        );
    }

    private isValidMessageType(type: string): boolean {
        return Boolean(
            config.has('order.messages.created') && (config.get('order.messages.created') as string[])?.includes(type),
        );
    }

    private hasExpectedMessageProperties(message: OrderCreatedMessage | OrderCustomerSetMessage) {
        return (
            !!(message as OrderCustomerSetMessage).customer ||
            (!!(message as OrderCreatedMessage).order &&
                (!!(message as OrderCreatedMessage).order?.customerEmail ||
                    !!(message as OrderCreatedMessage).order?.customerId) &&
                this.isValidState((message as OrderCreatedMessage).order?.orderState))
        );
    }
}
