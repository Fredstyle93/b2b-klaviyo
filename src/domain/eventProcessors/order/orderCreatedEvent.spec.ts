import { expect as exp } from 'chai';
import { DeepMockProxy, mockDeep } from 'jest-mock-extended';
import { OrderCreatedEvent } from './orderCreatedEvent';
import { MessageDeliveryPayload } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/subscription';
import { ctAuthNock, ctGetOrderByIdNock } from '../../../test/integration/nocks/commercetoolsNock';
import config from 'config';

const contextMock: DeepMockProxy<Context> = mockDeep<Context>();
contextMock.currencyService.convert.mockImplementation((value, currency) => value);

describe('orderCreatedEvent > isEventValid', () => {
    it('should return valid when is an Imported message', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'order' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: 'OrderImported' });
        Object.defineProperty(ctMessageMock, 'order', { value: { customerId: '123-1230-123', orderState: 'Open' } });

        const event = OrderCreatedEvent.instance(ctMessageMock, contextMock);

        exp(event.isEventValid()).to.be.true;
    });

    it('should return valid when is an OrderImported message', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'order' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: 'OrderCreated' }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'order', { value: { customerId: '123-1230-123', orderState: 'Open' } }); //mock readonly property

        const event = OrderCreatedEvent.instance(ctMessageMock, contextMock);

        exp(event.isEventValid()).to.be.true;
    });

    it.each`
        resource     | type                  | order                             | customer
        ${'invalid'} | ${'OrderCreated'}     | ${{ customerId: '123-1230-123' }} | ${null}
        ${'order'}   | ${'invalid'}          | ${{}}                             | ${null}
        ${'order'}   | ${'OrderCreated'}     | ${null}                           | ${null}
        ${'order'}   | ${'OrderCustomerSet'} | ${null}                           | ${null}
    `(
        'should return invalid when is not an orderCreated message, resource: $resource type: $type',
        ({ resource, type, order, customer }) => {
            const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
            Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: resource } }); //mock readonly property
            Object.defineProperty(ctMessageMock, 'type', { value: type }); //mock readonly property
            Object.defineProperty(ctMessageMock, 'order', { value: order }); //mock readonly property
            Object.defineProperty(ctMessageMock, 'customer', { value: customer }); //mock readonly property

            const event = OrderCreatedEvent.instance(ctMessageMock, contextMock);

            exp(event.isEventValid()).to.be.false;
        },
    );

    it('should return invalid when the orderState is invalid', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'order' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: 'OrderCreated' }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'order', {
            value: { customerId: '123-1230-123', orderState: 'FakeState' },
        }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'customer', { value: null }); //mock readonly property

        const event = OrderCreatedEvent.instance(ctMessageMock, contextMock);

        exp(event.isEventValid()).to.be.false;
    });
});

describe('orderCreatedEvent > generateKlaviyoEvent', () => {
    it('should generate the klaviyo event for an OrderCreated message', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'order' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: 'OrderCreated' }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'order', {
            value: {
                customerId: '123-123-123',
                customerEmail: 'test@klaviyo.com',
                orderState: 'Open',
                totalPrice: { type: 'centPrecision', centAmount: 1300, currencyCode: 'USD', fractionDigits: 2 },
                createdAt: '2023-01-27T15:00:00.000Z',
            },
        }); //mock readonly property

        const event = OrderCreatedEvent.instance(ctMessageMock, contextMock);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.eq(1);
        expect(contextMock.currencyService.convert).toBeCalledTimes(1);
        expect(contextMock.currencyService.convert).toBeCalledWith(13, 'USD');
        expect(klaviyoEvent[0].body).toMatchSnapshot();
    });

    it('should generate the klaviyo events for line items in an OrderCreated message', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'order' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: 'OrderCreated' }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'order', {
            value: {
                customerId: '123-123-123',
                customerEmail: 'test@klaviyo.com',
                orderState: 'Open',
                createdAt: '2023-01-27T15:00:00.000Z',
                totalPrice: { type: 'centPrecision', centAmount: 1300, currencyCode: 'USD', fractionDigits: 2 },
                lineItems: [
                    {
                        id: '123-123-123',
                        totalPrice: { type: 'centPrecision', centAmount: 1300, currencyCode: 'USD', fractionDigits: 2 },
                        name: {
                            en: 'Test product',
                        },
                    },
                ],
            },
        }); //mock readonly property

        const event = OrderCreatedEvent.instance(ctMessageMock, contextMock);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.eq(2);
        expect(contextMock.currencyService.convert).toBeCalledTimes(1);
        expect(contextMock.currencyService.convert).toBeCalledWith(13, 'USD');
        klaviyoEvent.forEach((event) => {
            expect(event.body).toMatchSnapshot();
        });
    });

    it('should generate the klaviyo event for an OrderCustomerSet message', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'order', id: '123123123' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: 'OrderCustomerSet' }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'customer', {
            value: {
                typeId: 'customer',
                id: '123123123',
            },
        }); //mock readonly property

        ctAuthNock();
        ctGetOrderByIdNock('123123123');

        const event = OrderCreatedEvent.instance(ctMessageMock, contextMock);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.eq(1);
        expect(contextMock.currencyService.convert).toBeCalledTimes(1);
        expect(contextMock.currencyService.convert).toBeCalledWith(13, 'USD');
        expect(klaviyoEvent[0].body).toMatchSnapshot();
    });
});
