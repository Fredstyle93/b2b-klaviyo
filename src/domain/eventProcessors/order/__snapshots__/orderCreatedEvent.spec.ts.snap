// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orderCreatedEvent > generateKlaviyoEvent should generate the klaviyo event for an OrderCreated message 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "Placed Order",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": undefined,
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`orderCreatedEvent > generateKlaviyoEvent should generate the klaviyo event for an OrderCustomerSet message 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "Placed Order",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customItems": [],
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "items": [],
        "lastModifiedAt": "2023-01-27T15:00:00.000Z",
        "orderId": "123123123",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "123123123",
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`orderCreatedEvent > generateKlaviyoEvent should generate the klaviyo events for line items in an OrderCreated message 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "Placed Order",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "items": [
          {
            "id": "123-123-123",
            "name": {
              "en": "Test product",
            },
            "totalPrice": {
              "centAmount": 1300,
              "currencyCode": "USD",
              "fractionDigits": 2,
              "type": "centPrecision",
            },
          },
        ],
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": undefined,
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`orderCreatedEvent > generateKlaviyoEvent should generate the klaviyo events for line items in an OrderCreated message 2`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "Ordered Product",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "id": "123-123-123",
        "name": {
          "en": "Test product",
        },
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "123-123-123",
      "value": 13,
    },
    "type": "event",
  },
}
`;
